File: 74-search-a-2d-matrix/search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        row, col = len(matrix), len(matrix[0])
        brow, frow = 0 , row - 1
        while brow <= frow:
            midrow = brow + ((frow - brow) // 2)
            if target > matrix[midrow][-1]:
                brow = midrow + 1
            elif target < matrix[midrow][0]:
                frow = midrow - 1
            else:
                break
        
        if not brow<=frow: return False

        bcol, fcol = 0, col -1
        mrow = brow + ((frow - brow) // 2)
        while bcol<=fcol:
            
            mcol = bcol + ((fcol - bcol) //2)
            if target > matrix[mrow][mcol]:
                bcol = mcol + 1
            elif target < matrix[mrow][mcol]:
                fcol = mcol - 1
            else: return True
        return False

File: 74-search-a-2d-matrix/search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int row = matrix.size();
        int col = matrix[0].size();

        int brow = 0;
        int erow = row - 1;
        while(brow <= erow){
            int mrow = brow + ((erow - brow)/ 2);
            if(target > matrix[mrow][col - 1]) brow = mrow + 1;
            else if (target < matrix[mrow][0]) erow = mrow - 1;
            else break;
        }

        if(brow>erow) return false;

        int bcol = 0;
        int ecol = col - 1;
        int mrow = brow + ((erow - brow)/ 2);

        while(bcol <= ecol){
            int mcol= bcol + ((ecol - bcol)/ 2);
            if(target > matrix[mrow][mcol]) bcol = mcol + 1;
            else if (target < matrix[mrow][mcol]) ecol = mcol - 1;
            else return true;
        }

        return false;
    }
};